### ISSUE TITLE
Windows build broken

### ISSUE THREAD

**Issue by petibub** [hoene/libmysofa#30]
The project for Windows got broken after the recent structural changes. I am unable to rebuild the Windows DLL. In particular, I was not able to solve the following problem: 

In mysofa.h: Cannot open include file: '../../build/src/mysofa_export.h': No such file or directory

I don't know in which directory this file is expected to be found, but it seems to be missing. Could you provide that file and clarify the directory?

Regards,
Piotr

**Comment by hoene** [comment]
currently, I am setting up an automatic compilation under Windows (with AppVeyor).
Give me some days, then it shall work again...

**Comment by petibub** [comment]
OK. Great. 

Unfortunately I don't understand what AppVeyor does. It seems like AppVeyor is targeting .NET applications. Their website has much information on how to use it, but is less informative on what is it doing :-(. 

Will the solution you're working on also work without AppVeyor and offer compiling to get a DLL?

**Comment by hoene** [comment]
yes, it will compile the dll for me.

**Comment by umlaeute** [comment]
@petibub `AppVeyor` is just a build service for native W32 builds (similar to Travis-CI which provides a un*x build-chain)

**Comment by petibub** [comment]
@umlaeute thanks for clarification. I'm not familiar with Travis-CI. CI is a cochlear implant, right? ;-). 

@hoene I was not able to create a functioning Windows version on the main branch (missing files) without reverting much of the job you've done. I tracked down the last functioning Windows version and tagged it as "v0.4(Windows)". In the code timeline, removal of pthreads was the next improvement of the main branch. I re-implemented it in the branch "Branch_v0.4". Thus, we have two branches now: the main (Linux development and the AppVeyor) and that one working in Visual Studio on Windows. The latest functioning version for Windows and VS is now tagged "v0.5(Windows)". 

Sorry for the mess with two branches - Is there a better way to proceed?

**Comment by umlaeute** [comment]
afaics, the problem is that @hoene's "fix" was to add some magic to the appveyor-script so that it actually succeeds.
since @petibub is not appveyor (at least: to his knowledge), he didn't do that magic.

the actual magic can be [found here](https://github.com/hoene/libmysofa/blob/master/.appveyor.yml#L6-L8) (so if you follow the "instruction" (read the source, luke), you should be able to get a proper build).

The actual build results (read: the pre-compiled library files) can be downloaded from the [appveyor project page](https://ci.appveyor.com/project/hoene/libmysofa) (go to a successful build and click on "Artifacts")

### here's a few thoughts to fix the mess
#### maintaining multiple build-systems
Maintaining multiple build systems is a pita. the main problem is obviously that two independent build-systems are bound to get out of sync.

Since `cmake` can generate `Visual Studio` solution files, why not use that instead? rather than maintaining old and half-maintained `.sln`/`.vc*proj` files (I'm just assuming "half-maintained" here for the sake of argument).

I *guess* that using `cmake` to generate the MSVC-Project file, will also take care of generating the `config.h` and similar files (so *no* magic is required).

what's the main obstacle to have `CMake` a hard build dependency for libmysofa?

#### copying files
independent of what i said above, i think there are better solutions for providing header files than overwriting the ones found in the repository with some system-specific files.
e.g.
- in the MSVC project file add a system specific Include directory, e.g. `windows\msvc\includes\`
  - make sure that this directory is included *first*
- move `windows\config.h.sln` to `windows\msvc\config.h`
- create `windows\msvc\mysofa_export.h` with something like:

      #include "mysofa_export.h.in"

this should allow one to use the current MSVC-project without any magic invocations beforehand.

having said that, we currently have the same information (`CPACK_PACKAGE_VERSION_*`) defined in multiple places, which makes it easy to forget changing them all simultaneously before a release.
it would be better if there was only a single place where the package-version gets defined.
which leads us back to "maintaining multiple build systems".

**Comment by umlaeute** [comment]
btw, the proper AppVeyor project page is currently: https://ci.appveyor.com/project/hoene/libmysofa-s142k

@hoene it seems there is no way to rename that project. however, if you delete both `libmysofa` projects and re-add one, it will show up as https://ci.appveyor.com/project/hoene/libmysofa (without the random hash at the end), which might be a nicer URL to remember :-)

**Comment by hoene** [comment]
@petibub 

Sorry for the mess I have made. 

Currently, we have to support four different OS environments including Debian, OSX, Windows (Visual Studio) and Windows (mingw).

Now, we have an automatic compilation for Debian, OSX and Windows (VS). 

Just go to 
https://github.com/hoene/libmysofa/commits/master

and click on the green check mark.

**Comment by umlaeute** [comment]
it also seems that recent versions of *Microsoft Visual Studio* support [opening CMake directly](https://blogs.msdn.microsoft.com/vcblog/2016/10/05/cmake-support-in-visual-studio/)

**Comment by hoene** [comment]
@umlaeute 
I tried to set up CMake to generate a Visual Studio project file. But I failed to achieve it (see https://ci.appveyor.com/project/hoene/libmysofa).
Now I am using Piotr MS project files (see https://ci.appveyor.com/project/hoene/libmysofa-s142k).

But thank you for the pointer, I will try to compile the CMake directly from Visual Studio...

**Comment by petibub** [comment]
Dear guys, 
thanks for the great explanations. I was not really into programming for years - I missed much of the development. 

@hoene I checked the green "check". Cool - it looks like appveyor can build the project. I will try to follow @umlaeute about copying files to get VS to compile the main branch outside appveyor. 

@umlaeute my first attempts to create a VS project was by opening the CMake in VS. Unfortunately, it created a complex project solution which was not even working. I had to clean it and remove most of the stuff. I probably don't understand CMake deep enough. But the solution I have now works and I'm fine. And when we get the main branch to compile for VS outside appveyor, I'll be more than happy. 

BTW, "half-maintained" is precise, indeed ;-).

**Comment by umlaeute** [comment]
@hoene i *think* I#ve gotten a bit further, but since I don not know how to use a zlib-installed nuget i'm stalled as well:
- https://ci.appveyor.com/project/umlaeute/libmysofa/
- with the appveyor.yml at: https://gist.github.com/umlaeute/4fefc4a75e3c5590c9e1712ddde09458

**Comment by petibub** [comment]
Guys,

I have now committed a version which can be compiled with VS without changing the sources. The only change was the solution: as a pre-build process, I mimic what this funny AppVeyor is doing, i.e., I let VS copy `mysofa_export.h.in` to `mysofa_export.h` and `config.h.sln` to `config.h`. 

So, now I don't need the additional Branch I've created. I hope that my solution does not break the AppVeyor flow.
